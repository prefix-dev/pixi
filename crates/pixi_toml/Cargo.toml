[package]
authors.workspace = true
description = "A crate that provides utilities to parse TOML files with extended diagnostics"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_toml"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
digest = { workspace = true }
hex = { workspace = true }
indexmap = { workspace = true }
itertools = { workspace = true }
miette = { workspace = true }
strsim = { workspace = true }
strum = { workspace = true }
toml-span = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
