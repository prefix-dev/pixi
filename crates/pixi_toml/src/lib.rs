//! This crate provides utilities to work with [`toml_span`].
//!
//! Deserialization helper structs are provide to deserialize types that are not
//! directly supported by [`toml_span`].
//!
//! This crate also provides [`TomlWith`] which enables directly parsing from
//! one type into another. This makes it possible to efficiently implement
//! deserializers for types from external crates.

mod diagnostic;
mod digest;
mod from_str;
mod hash_map;
mod index_map;
mod index_set;
mod one_or_many;
mod variant;
mod with;

use std::str::FromStr;

pub use diagnostic::TomlDiagnostic;
pub use digest::TomlDigest;
pub use from_str::TomlFromStr;
pub use hash_map::TomlHashMap;
pub use index_map::TomlIndexMap;
pub use index_set::TomlIndexSet;
pub use one_or_many::OneOrMany;
use toml_span::{DeserError, Spanned, Value, de_helpers::expected, value::ValueInner};
pub use variant::TomlEnum;
pub use with::TomlWith;

/// A trait that enables efficient deserialization of one type into another.
pub trait DeserializeAs<'de, T> {
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError>;
}

pub struct Same;

impl<'de, T: toml_span::Deserialize<'de>> DeserializeAs<'de, T> for Same {
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError> {
        T::deserialize(value)
    }
}

impl<'de, T, U: DeserializeAs<'de, T>> DeserializeAs<'de, Spanned<T>> for Spanned<U> {
    fn deserialize_as(value: &mut Value<'de>) -> Result<Spanned<T>, DeserError> {
        let span = value.span;
        let value = U::deserialize_as(value)?;
        Ok(Spanned { value, span })
    }
}

impl<'de, T, U> DeserializeAs<'de, Vec<T>> for Vec<U>
where
    U: DeserializeAs<'de, T>,
{
    fn deserialize_as(value: &mut Value<'de>) -> Result<Vec<T>, DeserError> {
        let array = match value.take() {
            ValueInner::Array(array) => array,
            other => {
                return Err(DeserError::from(expected("array", other, value.span)));
            }
        };

        let mut errors = DeserError { errors: Vec::new() };
        let mut result = Vec::with_capacity(array.len());
        for mut value in array {
            match U::deserialize_as(&mut value) {
                Ok(v) => result.push(v),
                Err(e) => errors.merge(e),
            }
        }

        if errors.errors.is_empty() {
            Ok(result)
        } else {
            Err(errors)
        }
    }
}

pub trait FromKey<'de>: Sized {
    type Err;

    fn from_key(key: toml_span::value::Key<'de>) -> Result<Self, Self::Err>;
}

impl<'de, T: FromStr> FromKey<'de> for T {
    type Err = <T as FromStr>::Err;

    fn from_key(key: toml_span::value::Key<'de>) -> Result<Self, Self::Err> {
        key.name.parse()
    }
}
