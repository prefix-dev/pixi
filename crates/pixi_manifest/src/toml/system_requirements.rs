use std::str::FromStr;

use pixi_toml::TomlFromStr;
use rattler_conda_types::Version;
use toml_span::{
    DeserError, Error, ErrorKind, Value,
    de_helpers::{TableHelper, expected},
    value::ValueInner,
};

use crate::{LibCSystemRequirement, SystemRequirements, system_requirements::LibCFamilyAndVersion};

impl<'de> toml_span::Deserialize<'de> for SystemRequirements {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        let mut th = TableHelper::new(value)?;

        let macos = th
            .optional::<TomlFromStr<_>>("macos")
            .map(TomlFromStr::into_inner);
        let linux = th
            .optional::<TomlFromStr<_>>("linux")
            .map(TomlFromStr::into_inner);
        let cuda = th
            .optional::<TomlFromStr<_>>("cuda")
            .map(TomlFromStr::into_inner);
        let libc = th.optional("libc");
        let archspec = th.optional("archspec");

        th.finalize(None)?;

        Ok(Self {
            macos,
            linux,
            cuda,
            archspec,
            libc,
        })
    }
}

impl<'de> toml_span::Deserialize<'de> for LibCSystemRequirement {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        match value.take() {
            ValueInner::String(str) => {
                let version = Version::from_str(&str).map_err(|e| Error {
                    kind: ErrorKind::Custom(e.to_string().into()),
                    span: value.span,
                    line_info: None,
                })?;
                Ok(LibCSystemRequirement::GlibC(version))
            }
            ValueInner::Table(table) => {
                let mut th = TableHelper::from((table, value.span));

                let family = th.optional("family");
                let version = th.required::<TomlFromStr<_>>("version")?.into_inner();

                th.finalize(None)?;
                Ok(LibCSystemRequirement::OtherFamily(LibCFamilyAndVersion {
                    family,
                    version,
                }))
            }
            inner => Err(expected("a string or table", inner, value.span).into()),
        }
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use rattler_virtual_packages::{Cuda, LibC, Linux, Osx, VirtualPackage};

    use super::*;
    use crate::{toml::FromTomlStr, utils::test_utils::format_parse_error};

    #[test]
    fn system_requirements_works() {
        let file_content = r#"
        linux = "5.11"
        cuda = "12.2"
        macos = "10.15"
        libc = { family = "glibc", version = "2.12" }
        "#;

        let system_requirements: SystemRequirements =
            SystemRequirements::from_toml_str(file_content).unwrap();

        let expected_requirements: Vec<VirtualPackage> = vec![
            VirtualPackage::Linux(Linux {
                version: Version::from_str("5.11").unwrap(),
            }),
            VirtualPackage::Cuda(Cuda {
                version: Version::from_str("12.2").unwrap(),
            }),
            VirtualPackage::Osx(Osx {
                version: Version::from_str("10.15").unwrap(),
            }),
            VirtualPackage::LibC(LibC {
                version: Version::from_str("2.12").unwrap(),
                family: "glibc".to_string(),
            }),
        ];

        assert_eq!(
            system_requirements.virtual_packages(),
            expected_requirements
        );
    }

    #[test]
    fn test_version_misspelled() {
        let input = r#"
        libc = { veion = "2.12" }
        "#;

        let result = SystemRequirements::from_toml_str(input).unwrap_err();
        assert_snapshot!(format_parse_error(input, result));
    }

    #[test]
    fn test_unknown_key() {
        let input = r#"
        lib = "2.12"
        "#;

        let result = SystemRequirements::from_toml_str(input).unwrap_err();
        assert_snapshot!(format_parse_error(input, result));
    }

    #[test]
    fn test_family_misspelled() {
        let input = r#"
        [libc]
        version = "2.12"
        fam = "glibc"
        "#;

        let result = SystemRequirements::from_toml_str(input).unwrap_err();
        assert_snapshot!(format_parse_error(input, result));
    }

    #[test]
    fn test_libc_misspelled() {
        let input = r#"
        [lic]
        version = "2.12"
        family = "glibc"
        "#;

        let result = SystemRequirements::from_toml_str(input).unwrap_err();
        assert_snapshot!(format_parse_error(input, result));
    }
}
