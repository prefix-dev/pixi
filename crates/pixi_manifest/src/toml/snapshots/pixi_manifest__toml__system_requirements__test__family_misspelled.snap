---
source: crates/pixi_manifest/src/toml/system_requirements.rs
expression: "format_parse_error(input, result)"
---
  × Unexpected keys, expected only 'family', 'version'
   ╭─[pixi.toml:4:9]
 3 │         version = "2.12"
 4 │         fam = "glibc"
   ·         ─┬─
   ·          ╰── 'fam' was not expected here
 5 │
   ╰────
  help: Did you mean 'family'?
